calyx/src/passes/infer_static_timing.rs

```diff
@@ -0,0 +1,71 @@
+use std::collections::HashMap;
+
+use crate::frontend::library::ast as lib;
+use crate::guard;
+use crate::ir;
+use crate::ir::traversal::{Action, Named, VisResult, Visitor};
+
+#[derive(Default)]
+pub struct InferStaticTiming {}
+
+impl Named for InferStaticTiming {
+    fn name() -> &'static str {
+        "infer-static-timing"
+    }
+
+    fn description() -> &'static str {
+        "infers and annotates static timing for groups when possible"
+    }
+}
+
+fn infer_cycles(dst_name: String, assignments: &Vec<ir::Assignment>) -> u64 {
+    for assign in assignments {
+        if assign.dst.borrow().get_parent_name() == dst_name {
```

Comment:
This seems like a pretty costly iteration, every time you call `infer_cycles`, you have to iterate over all assignments. I would have imagined that this can be one-pass backwards analysis where you start from the `done` hole and walk backwards.

Reply:
In my current implementation I'm still doing a lot of iterating over assignments, to find the `go` port corresponding to the `done` port that I'm computing the cycles-til-high number for.